typing/typeopt.ml

```diff
@@ -83,22 +83,23 @@ type classification =
   | Addr  (* anything except a float or a lazy *)
   | Any
 
-let classify env ty =
+let classify env ty : classification =
   let ty = scrape_ty env ty in
   if maybe_pointer_type env ty = Immediate then Int
   else match get_desc ty with
   | Tvar _ | Tunivar _ ->
       Any
   | Tconstr (p, _args, _abbrev) ->
-      if Path.same p Predef.path_float then Float
-      else if Path.same p Predef.path_lazy_t then Lazy
-      else if Path.same p Predef.path_string
-           || Path.same p Predef.path_bytes
-           || Path.same p Predef.path_array
-           || Path.same p Predef.path_nativeint
-           || Path.same p Predef.path_int32
-           || Path.same p Predef.path_int64 then Addr
-      else begin
+      begin match Predef.find_type_constr p with
+      | Some `Float -> Float
+      | Some `Lazy_t -> Lazy
+      | Some (`Int | `Char) -> Int
```

Comment:
This case is handled by the `maybe_pointer_type` call case, isn't it?

Reply:
Yes, this is correct. (We also know that the `data_type_constr` case below cannot be `Bool` for the same reason.) I would still write the code as-is, but maybe you had something else in mind?

(Maybe one should consider moving the check of `maybe_pointer_type` further down, for example in the `Type_abstract` case below, but I preferred to not do that to make the change less invasive and complex. The reason to do this would be a hope that, in the `int` or `char` case, checking `find_type_constr` first is very slightly more efficient. Not worth the ink to describe the change, I think.)